pkgs/build-support/fetchgit/nix-prefetch-git

```diff
@@ -148,7 +148,17 @@ checkout_hash(){
         hash=$(hash_from_ref "$ref")
     fi
 
-    clean_git fetch -t ${builder:+--progress} origin || return 1
+    if [[ -n "$deepClone" ]]; then
+        clean_git fetch -t ${builder:+--progress} origin || return 1
+        # Try to shallow fetch specific hash if it doesn't exist in any ref.
+        # For example, if a ref was force-pushed but the server still exposes the old hash.
+        clean_git fetch ${builder:+--progress} --depth 1 origin "$hash" 2> /dev/null
```

Comment:
I'm not sure if this fallback behavior is sound for `deepClone`, at least not in all cases.

`deepClone = true` usually implies `leaveDotGit = true` and most likely the `.git` folder would change between a shallow and full fetch changing the output hash right? This would mean that before and after a force push the `outputHash` would change and the derivation's output hash would no longer be reproducible this way. Making it fail by itself in that case is probably better UX?

Correct me if I'm wrong though.

Reply:
Hmm, that's a good point. `make_deterministic_repo` might keep the `.git` folder the same, but I'll test it out.

Reply:
If it doesn't, it would probably be better to update `make_deterministic_repo` so it does. A regular git clone would be able to fetch these commits, so nix-prefetch-git should also be able to fetch them without failing.